src/components/shared/helpers/normalizeProps.js

```diff
@@ -0,0 +1,35 @@
+const PRIMITIVE_TYPES = new Set(['string', 'number', 'boolean'])
+
+const createDefaultChangeHandler = onChange => payload => {
+    const value =
+        // ui-core event signature
+        (payload && payload.value) ||
+        // synthetic event
+        (payload && payload.target && payload.target.value) ||
+        // directly usable value
+        (PRIMITIVE_TYPES.has(typeof payload) && payload) ||
+        // ¯\_(ツ)_/¯
+        ''
+
+    onChange(value)
+}
+
+const normalizeProps = ({ input, meta, ...ownProps }, changeHandler) => {
```

Comment:
I get what's going on here, but I'm thinking if we could simplify this by breaking it up.

It's used for all components, and we're passing all the props through this function. That kind of obscures the fact that some props pass through unchanged, which you won't know from the caller unless you look at this function.

Breaking it up into more tightly scoped functions seems preferable to me. Though I know that is a vague recommendation and I'd have to take a closer look to see whether that's possible and how.

Reply:
We discussed to classify this as out-of-scope for now. I have created an issue so we don't forget to address this in the future: https://github.com/dhis2/ui-forms/issues/153